### ISSUE TITLE
[Request] Adding workflows rebuild openwrt like armbian

### ISSUE THREAD

**Issue by rapdodge** [ophub/amlogic-s9xxx-openwrt#268]
Dear ophub,

Can you make an workflows for github actions to rebuild openwrt using vanilla openwrt that downloaded like from https://downloads.openwrt.org/releases/21.02.3/targets/armvirt/64/ <- here

I'm requesting this, because i want to get stable openwrt (not snapshot) to make installing package that require kernel should be same is easier...

Ref: https://github.com/ophub/amlogic-s9xxx-openwrt/issues/241

Much thanks

**Comment by ophub** [comment]
https://github.com/ophub/amlogic-s9xxx-openwrt/tree/main/router-config#1011-required-options-for-openwrt


Since the official precompiled firmware lacks necessary dependencies, there are some problems in installation and use, and it cannot be used. You need to add these dependencies yourself to compile the appropriate crab

**Comment by rapdodge** [comment]
Noted, there is an imagebuilder official from openwrt to making rootfs

https://downloads.openwrt.org/releases/21.02.3/targets/armvirt/64/openwrt-imagebuilder-21.02.3-armvirt-64.Linux-x86_64.tar.xz

Can you please translating it to can be use github actions?

**Comment by ophub** [comment]
Implementing your needs is simple.
You send the official configuration, I will see if these dependencies are met, otherwise the firmware produced cannot be used normally.
Not everyone will read the documentation, and making firmware that doesn't work can cause a lot of problems.

**Comment by rapdodge** [comment]
Trying to get really vanilla openwrt

`wget https://downloads.openwrt.org/releases/21.02.3/targets/armvirt/64/openwrt-imagebuilder-21.02.3-armvirt-64.Linux-x86_64.tar.xz`

`tar -J -x -f openwrt-imagebuilder-21.02.3-armvirt-64.Linux-x86_64.tar.xz`

`cd openwrt-imagebuilder-21.02.3-armvirt-64.Linux-x86_64`

`make image PROFILE="Default" PACKAGES="\
perl-http-date perlbase-getopt perlbase-time perlbase-unicode perlbase-utf8 \
blkid fdisk lsblk parted attr btrfs-progs chattr dosfstools e2fsprogs f2fs-tools \
f2fsck lsattr mkf2fs xfs-fsck xfs-mkfs bsdtar pigz bash gawk getopt losetup pv tar \
uuidgen coremark coreutils coreutils-base64 coreutils-nohup kmod-brcmfmac kmod-brcmutil \
kmod-cfg80211 kmod-mac80211 hostapd-common wpa-cli wpad-basic iw" FILES="files"`

^[Ref](https://github.com/ophub/amlogic-s9xxx-openwrt/tree/main/router-config#1011-required-options-for-openwrt)

![image](https://user-images.githubusercontent.com/33513626/174315928-0d7873c9-d54c-477d-98a9-6519b55388ef.png)

Rootfs file https://transfer.sh/MOqJcr/openwrt-21.02.3-armvirt-64-default-rootfs.tar.gz

Github link: https://github.com/rapdodge/amlogic-s9xxx-openwrt/releases/tag/openwrt_s9xxx_official_06.17.2224

Above is just an barebone openwrt from [Ref](https://github.com/ophub/amlogic-s9xxx-openwrt/tree/main/router-config#1011-required-options-for-openwrt), but its nice to having other packages like luci, wireguard, etc

Will testing it right now...

**Comment by rapdodge** [comment]
Its on, and its working, STB can detect NIC plugged and unplug, but i cant setting for the internet, because i'm not familiar with the CLI...

![image](https://user-images.githubusercontent.com/33513626/174321540-855fc468-9c9e-4d3f-bc70-881806522e0d.png)

![image](https://user-images.githubusercontent.com/33513626/174321586-7b5f2524-4edf-41fd-a87b-d13f1b04233f.png)

![image](https://user-images.githubusercontent.com/33513626/174321615-aeb32f0b-84b7-4465-9a17-823cf4b8c7f3.png)

**Comment by ophub** [comment]
How to set up the network: ssh, `vi /etc/config/network`

**Comment by rapdodge** [comment]
> This is the first time I see this operation. Can I directly add dependency packages and plug-ins to the firmware? How long does it take you to append these dependencies?
> 
> Do you want one like this? https://github.com/ophub/amlogic-s9xxx-armbian/blob/main/.github/workflows/rebuild-armbian.yml
> 
> Download and compile directly from the given address?
> 
> ![Snip20220617_7](https://user-images.githubusercontent.com/68696949/174321675-b052775f-b126-4f66-8d60-108712065a61.png)

its not long, it just downloading it and configuring it...

Some screenshot
![image](https://user-images.githubusercontent.com/33513626/174322010-a1f63ae1-6b55-4079-9336-1594551689d1.png)

![image](https://user-images.githubusercontent.com/33513626/174322088-d7f6c02a-f116-4fb9-961b-e4c61c7e0888.png)

![image](https://user-images.githubusercontent.com/33513626/174322123-015bcf73-6f53-4e6d-915c-fb4ea3256125.png)

**Comment by rapdodge** [comment]
let me rebuilding it with luci... will keep you update

**Comment by ophub** [comment]
I asked you the question above, is the function you want the same as the one I posted?

**Comment by rapdodge** [comment]
> I asked you the question above, is the function you want the same as the one I posted?

yes, i'd like to have workflow like rebuild armbian :D

**Comment by ophub** [comment]
ok, very simple

**Comment by rapdodge** [comment]
Btw, if just rootfs without depedency that should be included on [ref](https://github.com/ophub/amlogic-s9xxx-openwrt/tree/main/router-config#1011-required-options-for-openwrt) if im not forgot, it cant be detect the NIC is plug or unplug...

I already doing this
https://github.com/rapdodge/amlogic-s9xxx-openwrt/blob/fd6781416d848c510da91fe809925227880d9dc6/.github/workflows/use-releases-file-to-packaging.yml#L112

**Comment by rapdodge** [comment]
let me try again... direct using openwrt default rootfs... but, before that, let me check using latest imagebuild with luci

**Comment by rapdodge** [comment]
Ok @ophub, i tested it with just openwrt base rootfs without customization, and it booting, but, NIC didnt respond...

**Comment by ophub** [comment]
This is the first time I have seen your compilation method. I used to build the tar.gz file completely from scratch. I have to learn to use this method, it seems very convenient, at least 3 hours of compilation time is omitted. If it's reliable, it's a great way.

**Comment by rapdodge** [comment]
Some other reference
https://openwrt.org/docs/guide-user/additional-software/imagebuilder

Looking forward into it :D

**Comment by zakeru46** [comment]
do you have amlogic service included? tried this imagebuilder method it seems need to be compiled first. I thought ophub's repack script has taken care of it.

**Comment by ophub** [comment]
The process of installing dependencies, does it support downloading third-party ipk installation?

**Comment by rapdodge** [comment]
Actually, I don't know...

**Comment by ophub** [comment]
yes

**Comment by rapdodge** [comment]
Ok, let me test it with my build openwrt barebone with luci, will update you asap

**Comment by rapdodge** [comment]
It's ok, no problem when installing it

![image](https://user-images.githubusercontent.com/33513626/174415004-89f801c5-1372-4229-ab45-5b04e8daa457.png)

![image](https://user-images.githubusercontent.com/33513626/174415028-fec04d13-875e-4772-813b-79e33e836093.png)

**Comment by ophub** [comment]
Can a third-party ipk be installed during imagebuild?

**Comment by rapdodge** [comment]
> Can a third-party ipk be installed during imagebuild?

I dont know, to be honest...

**Comment by ophub** [comment]
> And i think it more nice when the workflows can be added multi options what we want to build like just 21.02.3, 21.02.2, 21.02.1, etc...

> Because the link its likely same
https://downloads.openwrt.org/releases/ **version**/targets/armvirt/64/openwrt-imagebuilder...tar.xz


Send a few addresses like yours, let me see the rules, how to add a suitable drop-down list

**Comment by rapdodge** [comment]
Dear @ophub,

I think its 3rdparty ipk can be installed during imagebuild

Screenshot:

![image](https://user-images.githubusercontent.com/33513626/174416336-c1047211-7e5c-41b2-8cd4-d63f33cfdc63.png)

You should place custom ipk on packages folder inside the imagebuilder directory and adding it on make

![image](https://user-images.githubusercontent.com/33513626/174416388-e06f7103-3416-4011-a419-5c3e2e1a952f.png)

`make image PROFILE="Default" PACKAGES="\
perl-http-date perlbase-getopt perlbase-time perlbase-unicode perlbase-utf8 \
blkid fdisk lsblk parted attr btrfs-progs chattr dosfstools e2fsprogs f2fs-tools \
f2fsck lsattr mkf2fs xfs-fsck xfs-mkfs bsdtar pigz bash gawk getopt losetup pv tar \
uuidgen coremark coreutils coreutils-base64 coreutils-nohup kmod-brcmfmac kmod-brcmutil \
kmod-cfg80211 kmod-mac80211 hostapd-common wpa-cli wpad-basic iw \
liblucihttp liblucihttp-lua luci luci-app-firewall luci-app-opkg luci-base luci-lib-base \
luci-lib-ip luci-lib-jsonc luci-lib-nixio luci-mod-admin-full luci-mod-network \
luci-mod-status luci-mod-system luci-proto-ipv6 luci-proto-ppp \
luci-theme-bootstrap luci-compat luci-app-amlogic" FILES="files"`

**Comment by rapdodge** [comment]
> > And i think it more nice when the workflows can be added multi options what we want to build like just 21.02.3, 21.02.2, 21.02.1, etc...
> 
> > Because the link its likely same
> > https://downloads.openwrt.org/releases/ **version**/targets/armvirt/64/openwrt-imagebuilder...tar.xz
> 
> Send a few addresses like yours, let me see the rules, how to add a suitable drop-down list

dropdown i mean is like this

![image](https://user-images.githubusercontent.com/33513626/174416460-600e489f-f370-40be-b11e-61d44519f86b.png)

And adding latest major version like:
21.02.3 -> https://downloads.openwrt.org/releases/21.02.3/targets/armvirt/64/openwrt-imagebuilder-21.02.3-armvirt-64.Linux-x86_64.tar.xz

19.07.10 -> https://downloads.openwrt.org/releases/19.07.10/targets/armvirt/64/openwrt-imagebuilder-19.07.10-armvirt-64.Linux-x86_64.tar.xz

18.06.9 -> https://downloads.openwrt.org/releases/18.06.9/targets/armvirt/64/openwrt-imagebuilder-18.06.9-armvirt-64.Linux-x86_64.tar.xz

![image](https://user-images.githubusercontent.com/33513626/174416564-06789306-4b3d-41b3-9d36-7d4d7b27009e.png)

**Comment by rapdodge** [comment]
Testing using: https://github.com/rapdodge/amlogic-s9xxx-openwrt/releases/tag/openwrt_s9xxx_official_06.18.0914

It's booting: yes
It's can detect NIC: yes
It's have amlogic service: yes

**Comment by ophub** [comment]
https://github.com/ophub/amlogic-s9xxx-openwrt/tree/main/router-config/openwrt-imagebuilder

https://github.com/ophub/amlogic-s9xxx-openwrt/blob/main/.github/workflows/build-openwrt-with-imagebuilder.yml

You test, continue to add complete packages and luci-app, luci-theme

**Comment by rapdodge** [comment]
Noted, will testing it when I'm back home

**Comment by ophub** [comment]
https://github.com/ophub/amlogic-s9xxx-openwrt/blob/main/router-config/openwrt-imagebuilder/imagebuilder.sh

When using locally, just run `./imagebuilder.sh <branch>` directly, for example: `./imagebuilder.sh 21.02.3`

**Comment by rapdodge** [comment]
Building is under 20 mins
![image](https://user-images.githubusercontent.com/33513626/174437046-128328d6-15a7-4f87-89bc-db5947c96a8a.png)

Wonderful!

**Comment by rapdodge** [comment]
Testing: https://github.com/rapdodge/amlogic-s9xxx-openwrt/releases/tag/openwrt_s9xxx_imagebuilder_21.02.3_06.18.2012

It's booting: yes
It's can detect NIC: yes
It's have amlogic service: yes

Some screenshots

![192 168 1 1_cgi-bin_luci_admin_status_overview (1)](https://user-images.githubusercontent.com/33513626/174437862-90f294ab-75ba-41b2-b260-ddc2b48a15e3.png)

![image](https://user-images.githubusercontent.com/33513626/174437947-1a42bc05-52cb-4e56-ac17-1557ee07b110.png)

![image](https://user-images.githubusercontent.com/33513626/174437964-02e4ac24-1dc8-466e-877c-584a1e1c71fd.png)

This request i think can be closed

**Comment by rapdodge** [comment]
Thanks for your hardwork!
